### PR TITLE
Pilot: store latest push instead of global

### PR DESCRIPTION
**Pull request by howardjohn**
Attempt to fix https://github.com/istio/istio/issues/36791

### CODE REVIEW

**Comment by howardjohn** [comment]
cc @ramaraochavali @hzxuzhonghu WDYT? I cannot come up with any cases where it could break but want some more eyes

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -650,6 +653,7 @@ func (s *DiscoveryServer) computeProxyState(proxy *model.Proxy, request *model.P
 	if gateway && proxy.Type == model.Router {
 		proxy.SetGatewaysForProxy(push)
 	}
+	proxy.LastPushContext = push
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/ads.go:656]: Can you move this line up?

	if push == nil {
		push = s.globalPushContext()
                proxy.LastPushContext = push
	}

**Review by hzxuzhonghu — COMMENTED** [review]
LGTM

**Comment by ramaraochavali** [comment]
Where we using globalContext() now after this change? Probably in some tests only? Should we cleanup that also?

**Comment by howardjohn** [comment]
We use it when we initiate Pushes. Let me add comments and clean things up

**Comment by howardjohn** [comment]
> Where we using globalContext() now after this change? Probably in some tests only? Should we cleanup that also?

Cleaned up the remaining incorrect usages (there are a small number of correct usages) and added comments

**Review by ramaraochavali — APPROVED** [review]
Thankyou.

**Pull request merged**